MetaMask/metamask-mobile
# [5.1.1] Download attachments in browser on iOS

## What?
resolves https://github.com/MetaMask/mobile-planning/issues/186
This PR is dependant on [this change in the dapp repo](https://github.com/MetaMask/dapps/pull/121)

This PR adds the ability to download attachments from the MetaKask mobile browser on iOS. This is not needed on android because this functionality id built into the underlying implementation. I achieved this with the following changes...

- adding custom logic to the [React native webview onFileDownload prop](https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#onfiledownload)
- creating a [util function that downloads an attachment and then prompts the user to save](https://github.com/MetaMask/metamask-mobile/pull/4052/files#diff-82c166dfc5ec7e5b7d82875c105fa80055d89e6a377427f45d28df4379c75f88R22) it in the desired location and allows for renaming of the file.
- Refactoring the URL modal logic into its [own component](https://github.com/MetaMask/metamask-mobile/pull/4052/files#diff-6526c28c56229fd13f11760fa232ecb198b7b75efd48396c347d8f100436557cR15) that handles its own state. 
    - this component leverages the ReusableModal (courtesy of @Cal-L ❤️ ) which is no longer a native modal that competes with the native share modal. Without this the share modal will open then disappear right away since the URL modal overrides it. 
    - this also gives us some nice gesture handling now.
- defining the modals as modals in the [stack navigator](https://github.com/MetaMask/metamask-mobile/pull/4052/files#diff-b7701661b3e11676aba508c3ddaf1973fc5c1c6cac403c48408083c5645c3a0bR97-R109) so we don't need to conditionally render them in the view. This means we can treat modals the same way we treat normal navigations but they will appear as overlays to the screen.


https://user-images.githubusercontent.com/22918444/164080501-f65c1fee-22b4-4ac6-97ee-4889187214cc.mov


Here are some sample attachment urls
- https://storage.opensea.io/files/7c2a3bf635a404d6ad0cdbea5a643d6f.glb
- https://openseauserdata.com/files/7c2a3bf635a404d6ad0cdbea5a643d6f.glb

## How to test
- This is somewhat complex to test as you will need to spin up a local version of the [dapp site](https://github.com/MetaMask/dapps/pull/121)
1. pull the `blur-autocomplete` in the dapp repo from
2. run `yarn start` in the dapp repo which should open the https://home.metamask.io/ in http://localhost:3000/
3. open xcode and run this branch in on the simulator 
4. open the browser inside the mobile app and navigate to http://localhost:3000/
5. paste one of the test links I provided above into the search bar
6. the download should trigger and the save file modal should pop up letting you rename the file and decide where to save the file on the device. 

### ⚠️  Note about testing with the simulator
while most of these features are testable on the simulator and you `need to run on the simulator to test the new dapp` functionality, some behaviour has been weird.
- I have not been able to actually view the files in the file system on the simulator but it is working on my real device
To test on a real device just ignore the part about spinning up a local version of the Metamask home page and simply click the top address bar to make your searches. 

### Expected behaviour
- [ ] clicking the top address bar on any page should open the url modal
- [ ] you should be able to cancel OR swipe the modal down
- [ ] when the modal opens `on the home page`, there should be no pre populated text in the search field
- [ ] when the modal opens `on any page other than the home page`, the current url should be in the pre populated text
- [ ] searching urls brings you to the exact url
- [ ] searching terms brings you to a duck duck go search result
- [ ] when you navigate to a url that is an attachment (determined by the Content-Disposition header) it triggers a download and prompts you to save the file
- [ ] after saving the file you can see it in your file system of your ios device under the location and name you saved it

**Checklist**


**Issue**
resolves https://github.com/MetaMask/mobile-planning/issues/186
